akka-http-core/src/main/scala/akka/http/Http.scala

```diff
@@ -173,6 +174,217 @@ class HttpExt(config: Config)(implicit system: ActorSystem) extends akka.actor.E
+    }
+  }
+
+  /**
+   * Fires a single [[HttpRequest]] across the (potentially cached) host connection pool for the request's
+   * effective URI to produce a response future.
+   *
+   * Note that the request must have either an absolute URI or a valid `Host` header, otherwise
+   * the future will be completed with an error.
+   */
+  def singleRequest(request: HttpRequest,
+                    options: immutable.Traversable[Inet.SocketOption] = Nil,
+                    settings: HostConnectionPoolSettings = HostConnectionPoolSettings(system),
+                    defaultHeaders: immutable.Seq[HttpHeader] = Nil,
+                    log: LoggingAdapter = system.log)(implicit fm: FlowMaterializer): Future[HttpResponse] =
+    try {
+      val setup = ConnectionPoolSetup(encrypted = false, options, settings, defaultHeaders, log)
+      val effectiveRequest = request.withEffectiveUri(securedConnection = false)
+      val uri = effectiveRequest.uri
+      val gateway = cachedGateway(HostConnectionPoolSetup(uri.authority.host.toString(), uri.effectivePort, setup))
+      val responsePromise = Promise[HttpResponse]()
+      gateway.onComplete(dispatch(request, responsePromise))(system.dispatcher)
+      responsePromise.future
+    } catch {
+      case e: IllegalUriException ⇒ FastFuture.failed(e)
+    }
+
+  /**
+   * Triggers an orderly shutdown of all host connections pools currently maintained by the [[ActorSystem]].
+   * The returned future is completed when all pools that were live at the time of this method call
+   * have completed their shutdown process.
+   *
+   * If existing pool client flows are re-used or new ones materialized concurrently with or after this
+   * method call the respective connection pools will be restarted and not contribute to the returned future.
+   */
+  def shutdownAllConnectionPools(): Future[Unit] = {
+    import scala.collection.JavaConverters._
+    import system.dispatcher
+    val gateways = hostPoolCache.values().asScala
+    system.log.info("Initiating orderly shutdown of all active host connections pools (w/ configured grace period)...")
+    Future.sequence(gateways.map(_.flatMap(_.shutdown()))).map(_ ⇒ ())
+  }
+
+  private def cachedGateway(setup: HostConnectionPoolSetup)(
+    implicit fm: FlowMaterializer): Future[ConnectionPool.Gateway] = {
+    val promise = Promise[ConnectionPool.Gateway]()
+    hostPoolCache.putIfAbsent(setup, promise.future) match {
+      case null ⇒ // we are the first to attempt starting a pool for this setup
+        try {
+          val gateway = ConnectionPool.startHostConnectionPool(setup)
+          promise.success(gateway)
+          gateway.whenShuttingDown.onComplete(_ ⇒ hostPoolCache.remove(setup))(system.dispatcher)
+          promise.future
+        } catch {
+          case NonFatal(e) ⇒
+            promise.failure(e)
+            hostPoolCache.remove(setup) // remove Cowboy entry if we weren't able to properly start the pool
```

Comment:
Nope. No more cowboys here. This is a civilized place now.
I'll remove the signpost to the former wild-west territory.